cmd/deploy.go

```diff
@@ -95,192 +88,312 @@ func postDeploymentRequest(l log.Logger, projectName string, namespaceName strin
+			if !resp.GetSuccess() {
+				streamErrs = append(streamErrs, errors.New(resp.GetMessage()))
 			}
-			adaptedJobSpecs = append(adaptedJobSpecs, adaptJob)
+			counter++
+			spinner.SetProgress(counter)
 		}
-		respStream, err := runtime.DeployJobSpecification(deployTimeoutCtx, &pb.DeployJobSpecificationRequest{
-			Jobs:          adaptedJobSpecs,
-			ProjectName:   projectSpec.Name,
-			NamespaceName: namespaceSpec.Name,
-		})
-		if err != nil {
-			if errors.Is(err, context.DeadlineExceeded) {
-				l.Error(coloredError("Deployment process took too long, timing out"))
-			}
-			return fmt.Errorf("deployement failed: %w", err)
+	}
+	spinner.Stop()
+	if len(streamErrs) > 0 {
+		for _, e := range streamErrs {
+			fmt.Println(e)
 		}
+		return errors.New("one or more errors are encountered during deploy")
+	}
+	return nil
+}
 
-		ackCounter := 0
-		totalJobs := len(jobSpecs)
-		spinner := NewProgressBar()
-		if !verbose {
-			spinner.StartProgress(totalJobs, "please wait")
-			spinner.SetProgress(0)
-		}
-		var streamError error
-		for {
-			resp, err := respStream.Recv()
-			if err != nil {
-				if err == io.EOF {
-					break
-				}
-				streamError = err
-				break
-			}
-			if resp.Ack {
-				// ack for the job spec
-				if !resp.GetSuccess() {
-					return fmt.Errorf("unable to deploy: %s %s", resp.GetJobName(), resp.GetMessage())
-				}
-				ackCounter++
-				spinner.SetProgress(ackCounter)
-				if verbose {
-					l.Info(fmt.Sprintf("%d/%d. %s successfully deployed", ackCounter, totalJobs, resp.GetJobName()))
-				}
-			} else {
-				if verbose {
-					// ordinary progress event
-					if resp.GetJobName() != "" {
-						l.Info(fmt.Sprintf("info '%s': %s", resp.GetJobName(), resp.GetMessage()))
-					} else {
-						l.Info(fmt.Sprintf("info: %s", resp.GetMessage()))
-					}
-				}
-			}
+func registerAllNamespaces(
+	deployTimeoutCtx context.Context, runtime pb.RuntimeServiceClient,
+	l log.Logger, conf config.Optimus, namespaceNames []string,
+) error {
+	var selectedNamespaceNames []string
+	if len(namespaceNames) > 0 {
+		selectedNamespaceNames = namespaceNames
+	} else {
+		for _, namespace := range conf.Namespaces {
+			selectedNamespaceNames = append(selectedNamespaceNames, namespace.Name)
 		}
-		spinner.Stop()
+	}
 
-		if streamError != nil {
-			if ackCounter == totalJobs {
-				// if we have uploaded all jobs successfully, further steps in pipeline
-				// should not cause errors to fail and should end with warnings if any.
-				l.Warn(coloredNotice("jobs deployed with warning"), "err", streamError)
-			} else {
-				return fmt.Errorf("failed to receive success deployment ack: %w", streamError)
-			}
+	ch := make(chan error, len(selectedNamespaceNames))
+	defer close(ch)
+	for _, namespaceName := range selectedNamespaceNames {
+		go func(name string) {
+			ch <- registerNamespace(deployTimeoutCtx, runtime, l, conf, name)
+		}(namespaceName)
+	}
+	var errMsg string
+	for i := 0; i < len(selectedNamespaceNames); i++ {
```

Comment:
Can range over the channel: https://gobyexample.com/range-over-channels

Reply:
if we range from the channel, at some point we will hit a deadlock error caused by the goroutine not sending anything to the channel, that's why we use this instead